### ISSUE TITLE
[UX] Display a notification badge in the admin bar when there are warnings/errors.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#639]
I really hate it to visit the site status report only to find that there are errors/warning that were sitting there for some time waiting for action (db updates for one).

So, how about we have some discrete badges showing up over the top-right corner of the "Reports" top-level menu? They could be as discrete as a simple red/orange dot (depending on the highest severity of the errors/warnings) similarly to the way that git hub does it:

![github_notifications_badge](https://cloud.githubusercontent.com/assets/2423362/5846363/20575f16-a217-11e4-8327-e30c4445580b.png)

Some popup note would be nice too:

![github_notifications_popup](https://cloud.githubusercontent.com/assets/2423362/5846385/5246b3dc-a217-11e4-9cc6-065df0f49034.png)

If there is only one error/warning, then the popup message could be populated with a specific text like "Database updates: Out of date" or "HTTP request status: Fails" or "Cron maintenance tasks: Last run [long time] ago". If there are multiple errors/warnings, then something generic like "You have unresolved issues with your site".

**Comment by chrisherberte** [comment]
:+1:

**Comment by quicksketch** [comment]
This is a great idea.

**Comment by BWPanda** [comment]
I agree, with the request that we allow issues/errors to be ignored so the notification doesn't always show for something that can't/won't be fixed (e.g. 'HTTP request status' or 'Out of date').

**Comment by sutibun** [comment]
Notification Center: Because a badge is not enough
====

When I read about the notification badge, I didn't think too much of it. Thought it was fine. Probably because I was [leaning another way](https://github.com/backdrop/backdrop-issues/issues/495#issuecomment-139957730). What inspired and changed my thinking was experiencing the badge recently while using gmail.

Long story short, I've changed my mind and now this has my full support. **Not only do I think Backdrop needs this,  but I think it should be bigger.** From his description, I'm not sure if @klonos had this in mind or probably didn't have enough time to give the details.

Either way, I'd like to make it explicit here.

What's the problem?
---

Let's start from the beginning.

My problem lies in module installation and configuration. There are two problematic scenarios I've come across consistently in Drupal/Backdrop. One with core and the other with contrib modules.

1. **When you upgrade Backdrop to a newer version, that means the possibility of new core modules/new features added. (e.g. Rich text editor) The issue is how will users learn about them?** Not everyone will keep up to date with latest Bd development. (I don't keep up with what's new in Wordpress. I just click upgrade without much thought.)
2. **When you enable a contrib module, new users won't necessarily know how to configure them or if s/he needs to do anything else special.** They'll think enabling is all.
    - I'm not sure if this happened to anyone but I've installed some new modules and forgot the name of it. I've had to scroll down and figure what was new.
    - Yes, there is a config/edit/permission link in the modules listing but after enabling it, I've had to find it by scrolling down or searching. This extra work seems unnecessary.
    - I remember a Drupal module (think the AES module) that after enabling displayed a message above the module list. It provided a message saying what you needed to do to get it to work, along with a link to the exact page to configure it. It was *extremely* helpful. Many Wordpress plugins do this in fact, and so I am never lost. These messages are like essential guides to new users. I like it because I don't have to look around or wonder. The module tells me exactly. Drupal veterans already know the drill so they might forget that it isn't always obvious. With Drupal/Bd, I am on my own after enabling modules.

The first issue is easily solved with a [Welcome page which notifies users with what's new in this version](https://github.com/backdrop/backdrop-issues/issues/1225#issuecomment-141605031).

The only issue left is configuring modules and doing extra necessary work.

Possible solutions
----

You could 

- add a message on the status report that you need to properly configure a module. However, in the time I've used Drupal, I've not been trained to go there, except maybe when there's a new Drupal update. The status report is not the first place I think of. You still run into the problem of notifying the user that there's something wrong. After I enabled the CKEditor core module, for example, I had no idea I had go to text formats.
- add a message on top of the page but then if you click away by accident, it goes away. If you make the message permanent, it'll follow you everywhere until you dismiss it. Plus, you can get a whole pile of messages. Admittedly, it's not bad but there's a better way.


Backdrop needs to centralize all notifications
-----

What inspired me was when I received a notification popup on GMail. Since for some reason I'm thinking Backdrop in my sleep, I was reminded of @klonos' badge idea.

![google](https://cloud.githubusercontent.com/assets/14132746/10120668/5fb37184-648d-11e5-9cb3-dde5f1f90f24.png)

That was it though. Just a reminder.

Then when using an iOS device, I noticed the lock screen had the "notification center" with messages. i.e. a central place where apps can notify users. And this got me thinking.

What if the Backdrop CMS was like the Apple platform and modules are apps. Because Bd is a platform, it needs to provide organization and infrastructure so everything meshes well together. **Just like iOS notifications are centralized, so should Backdrop notifications.**

Here is what I have in mind on how this will work.

### Notification inbox

Any and all notifications on Bd must use the Notification inbox as this will be the way to notify users of anything. The centralized place.

![indicator](https://cloud.githubusercontent.com/assets/14132746/10120669/6388ed48-648d-11e5-9c5f-8084de8d5033.png)

1. Whenever a module needs to notify the user, it pushes its message to the Bd notification inbox.
2. Upon receiving a new message, the badge will display a marker/badge alerting of a new notification.
3. The user will see this and click on it to see what's up.
4. A pop up will appear similar to the the GMail one, displaying the message.
5. Once the message is read, the badge/marker disappears.

For critical/urgent messages, they'll remain as a sticky on the popup (maybe change marker color also to designate degree of importance). Useful for Bd security upgrades. 

### Notification/Message center

However, if you accidentally dismiss a message or just want to revisit an old message, there will be a link to the message center. The message center will be like an archive page, listing past messages. If by chance you install a lot of modules, you won't lose the messages.

![notification-center](https://cloud.githubusercontent.com/assets/14132746/10120671/6723769e-648d-11e5-9e83-76f8da6c6364.png)


The message center won't list everything in one list. It'll have different panels. The message center page should be organized so important messages are the focus. Everything else should be accessible and visible but to the side. (Can probably create different urgency labels)

- Messages can be categorized with a given priority. from contrib, core, warning,
- Removes the need for permanent messages floating around and pushing content down. Or accidentally clicking away and not seeing the temp message.

The long term vision
-----

> So, all this for module notifications?

At first, yes. Let's start simple. Long term, no.

Open your imagination. Now that we have a centralized place to notify users, what possibilities does this open?

- What if bdcms.org pushed tailored warnings only to site admins with a certain setup or an insecure module. Or, maybe notify selected admins that a discovered bug affects their system with a link to the GitHub issue.
- Notify of a new Bd upgrade.
- What if an editor is notified to revise the draft a writer has just submitted.
- What if admin is notified of a new user waiting for their account to be activated.
- What if there are notifications for new comments?
- What if there was a warning when significant log in attempts have been made within a certain time frame?
- possibility of notifying and inviting admins to Backdrop events near their area
- Possibility of notifying all Bd admins that an important vote is being called on an issue that affects the community and they need feedback.
- What if after installing and enabling any module, a notification appears with a successful message and a link to its configuration page (for those modules that don't provide anything custom) No need to search or scroll down. The config links are within easy reach.

A lot of these notification can be changed based on user permissions and can be configured so it doesn't get annoying.

\*wipes forehead\* Ok. Monkey is done. I am happy I finally got this off my chest.

**Comment by quicksketch** [comment]
Thanks @sutibun for the impressive summary! It would be very nice to solve these problems in Backdrop. You've got me thinking about it and just as a mental exercise, here's a technical run-down on how I would hope we'd tackle this:

- This would almost certainly need to be a new module, e.g. `notifications.module`. As it would need everything: dedicated menu items, configuration, a default layout (e.g. a notifications dashboard), Views integration, alter hooks for modifications to the admin bar, etc.
- Notifications would probably be a new entity type, because they likely need to be translated. (Although translation UI for entities isn't comprehensive yet.) We'd then use Views to display each block of notifications.
- We'd run a cron job to pull updates from BackdropCMS.org, creating new and deleting old notifications as needed. As this would be a continuous thread, we'd want to avoid bloating the notifications database table just with our own announcements. Of course this would need to be configurable to disable notifications entirely from BackdropCMS.org, in which case the block/view for these notifications should be disabled.
- Because the long-term uses of notifications as brought up (e.g. new comment notifications, private messages, etc), this functionality is likely to be considered *non-administrative`. That is, the path for notifications would likely be `/notifications` rather than `/admin/notifications`, because non-admin users may also have notifications. But like `node/add`, we could specify this path as administrative by default, so that it used the admin theme for users that had access to it.

One thing that won't be clear without a lot more thinking is how access control should work on these notifications. Some of the examples mentioned are clearly per-user, others are global. We'd need a way to query (via views) the notifications that are appropriate for any particular user, and then thread them together. That problem is not an easy one. Nodes have access control through a rather complicated [`hook_node_access_grants()`](https://api.backdropcms.org/api/backdrop/core!modules!node!node.api.php/function/hook_node_access_records/1) and hook_node_access(). I think everyone would like a simpler solution for notifications, like maybe just a global "view administration notifications" permission.

Anyway, I think this sounds pretty great, but also really ambitious (which I also like). But the scope of this is so substantial, it's going to take months of work to get worked out. By comparison, we could probably get the notification badge alone into the admin bar easily within a few days/weeks once we start development. Should we start small (notification badge for status report) and then move to the more comprehensive solution?

**Comment by sutibun** [comment]
hehe. I didn't meant to imply this should get done all at once. That whole post was more like a big dump of my vision, a destination map of sorts. I meant to include a plan of action at the end but I got tired and sleepy. I just needed to get the big picture out of my head and express it before it continued bothering me. I feel like a huge weight lift has been lifted, btw.

Maybe I am simple but my philosophy is to

- start small
- go for what's practical, something we can use now (or near future) Not possible future problems
- reuse what we already have like Layouts, Views, fields, pages, etc. I think if a module is in core, they're meant to be used as fundamental blocks to build other system parts (i.e. core modules = lego pieces = atoms)

My plan of action is similar to what I said in the [home page layout](https://github.com/backdrop/backdrop-issues/issues/1042#issuecomment-142806078)
> I say let's start simple and build off a basic foundation. We don't need to build anything complex immediately.

What I had in mind was 

- [ ] **get the notification inbox onto the admin bar**
- [ ] **restrict notifications to core modules for now** (until we figure out how to bring in contrib modules)
    - I didn't even consider translations. Earlier today I also didn't think that a contrib module may start spamming users. This is why core modules notifications should be perfected first.
- [ ] **a ridiculously simple notification center page** (basically a giant listing of notifications)
    - Even this, we might not even need right away. It could just be the popover with a button to clear each message.
- [ ] **whenever you install a new module, a notification appears linking to any config + permission pages the contrib provides.** (with confirmation, "Success You've installed the X module. Take a look here to configure it. And here to set permissions")
    - But we can go for the notification badge for the status report instead

**Comment by quicksketch** [comment]
Hi guys, well for starters I wonder if a notification badge linking to the Status Report would be a starting point? I hacked this together to see the effect:

![selection_010](https://cloud.githubusercontent.com/assets/100206/10153877/ff17f45c-661d-11e5-92b5-893fa78f7d42.png)

But honestly it's just so much less than what has been stated here. This is purely a badge that modules can use `hook_admin_menu_build_alter()` to add if they want to mark their menu items as having a notification. What do you think, too small to implement?

**Comment by klonos** [comment]
> What do you think, too small to implement?

Good start and better than nothing. One thing though: how do we get the users' attention to second-level items? Do we show the badge in each level, all the way down to the item in question?

**Comment by klonos** [comment]
...also, does it make sense to have these colored as red or orange depending on whether it is a warning or error? What if multiple, mixed warnings and errors exist?

**Comment by quicksketch** [comment]
Just to be clear, I just did this because this is a "couple hours" job. Even the basic implementation described is still months of work. Just having a simple notifications page doesn't mean that we wouldn't need an awful lot of infrastructure to get to that point. We can't add a mailbox icon if we don't have a notifications page to which it should be linked. And we can't a notifications page without a notifications data storage of some kind.

> One thing though: how do we get the users' attention to second-level items?

This is a great question. I'm not sure.

> Do we show the badge in each level, all the way down to the item in question?

I would expect to be able to click on the item that had the notification badge, so adding it to the top level item might cause some usability problems. If I saw a badge next to "Reports", I'd click on it directly.

> ...also, does it make sense to have these colored as red or orange depending on whether it is a warning or error? What if multiple, mixed warnings and errors exist?

I was thinking that we'd only display the indicator for highest level problem. So if there were 2 warnings but one error, you'd just get a red `1`. After resolving that issue, you'd get a yellow `2` in the same place.

**Comment by sutibun** [comment]
For starters, it's not bad, especially if to alert you to issues in the status report and maybe a few other areas that makes sense (e.g. 10 new comments) I like the red on black look. The main weakness of this approach is you're relying on users hovering over a *particular* parent menu in order to see the red badge under it. **It's a passive alert.**

If the status report issue is urgent, there's a chance the user will miss it. To go around this, you could add a badge on the parent menu as well.

**But I see it as a short term solution**, as a way to get *something* out now instead of waiting months. If that's the case, that's not a problem. Practicality has to win.

Long term, I don't suggest this approach because

- Individual modules adding a badge on their menu item can clutter the admin menu. You could have something like the red badges on iPhone homescreens
    
    ![iphone-badge](https://cloud.githubusercontent.com/assets/14132746/10155567/875a84ba-663c-11e5-8229-49e59c89a54b.png)
    Which one is most important? Which one deserves your attention first?
- If you see red badges everywhere in the admin menu, none of them stand out and in time you begin to ignore them. The opposite of what you'd want.
- You'll create a new habit where users hover over the admin menus to see if there's any new alerts (think of checking email every minute)

Something that came to mind is [implementing a scroll trigger box](http://codecanyon.net/item/scroll-triggered-box-for-drupal/full_screen_preview/10853696) (Just scroll all the way down). Basically, a message pops up on the bottom right corner of the page when you scroll past a certain point. It certainly grabs attention.

![scroll-trigger](https://cloud.githubusercontent.com/assets/14132746/10155783/c60c8bde-663e-11e5-8577-9e2edd7f9241.png)

Maybe you could do that but modify it so it appears when you want it to, instead of it being dependent on scrolling. If there's an issue in the status report, the box appears on the bottom right corner. Once you take action, it clears away.

I don't know.

I tried to think of alternatives but kept returning to the following approach if you want something quick to implement. Just use what Drupal already uses: a message that appears on top. It's the most practical solution.

![drupal-message](https://cloud.githubusercontent.com/assets/14132746/10155390/2f6fef08-663a-11e5-9c28-d06faf689585.png)

- The message pops up when you want it to
- If the user doesn't need it anymore, it goes away
- The message appears in the same place over and over so it becomes familiar
- The message can easily be customized and be as long as you want it

Maybe what you can add is make the messages follow you around until you manually clear it. Maybe pretty it up.

**The key to a good notification, IMO, is grabbing user attention when you need it and making sure you train users to expect it in one area, as opposed to multiple places.**

This message on top achieves both. Lots of WP plugins use this so its effective. Since it's already built into Drupal, there's "little" work to do.

In other words, I personally like the notification inbox idea but if it's impractical, I recommend the approach where messages appear on top. Much simpler.

**Comment by jenlampton** [comment]
@quicksketch can we add the red dot to the top-level Reports item as well as Status report?  I think that would solve the short-term problem and might be easy enough to get into 1.3.0

**Comment by docwilmot** [comment]
An initial concept here:

State when there is a warning (it would be red if there is an error):
![alert button](https://cloud.githubusercontent.com/assets/1042283/12223886/3b305f08-b7b1-11e5-8cc5-f4b8f315ca95.PNG)

Hover state:
![hover](https://cloud.githubusercontent.com/assets/1042283/12223893/5ee0ab74-b7b1-11e5-84f0-b5a2a3933d8c.png)

This implementation does a few things:
- implements a new hook: `hook_admin_bar_alerts()` which modules can use to let Admin Bar know there are errors/warnings
- system module implements this by default
- so far, havent implemented a means to make the icon dismissable
- icon is CSS based and can be altered (for now just a `background-color:red`, but could be a different image for warning vs error.
- hovering over the icon gives a list of all errors provided by modules
- modules can add a url to go to to fix the error or find more info

Note!: I made the Admin Bar render on demand rather than caching the rendered menu so this is easier to test. This is not meant to be permanent.

Proof of concept for now.

**Comment by quicksketch** [comment]
Thanks @docwilmot. I think both my first approach and this approach will need to figure out some kind of smarter way to cache/expire/dismiss these warnings. The entire admin bar is cached (a very good thing) and adding a dynamic bit to it like this means that we'll either need to expire the cache after changes (such as dismissing or resolving an issue), or load in this dynamic part after the initial page load.

> implements a new hook: hook_admin_bar_alerts() which modules can use to let Admin Bar know there are errors/warnings

First of all, I like this a lot better than using a hack like mine (using `hook_admin_menu_build_alter()`).

Should this actually be namespaced within admin bar? Perhaps it should be a system.module provided concept with a name like `hook_system_alerts()` or even just `hook_alerts()` (though namespacing could be an issue there).

**Comment by docwilmot** [comment]
I hadn't noticed your idea, @quicksketch. I do agree this would need a lot more thinking. Only attempted it because it's tagged for 1.3.0.

**Comment by docwilmot** [comment]
Please re-tag if you agree this should be postponed.

**Comment by jenlampton** [comment]
I think we can definitely make something that's ready for 1.3.0, and is still a sifnificant improvement over what we have now.  Proposal:
- we add an indicator in the admin bar that is a link to the status report page.
- we remove that horrible backdrop_set message that appears on every admin page, that provided a link to the status report page.

All we're really doing now is moving the alert from something that was annoying and only on half the pages to something that's helpful and on every page of the site. The rest we rethink and add later on :)

@docwilmot I think your PR is a great start!

**Comment by jenlampton** [comment]
This new PR gets rid of the `backdrop_set_message` at the top of the page, and instead adds a red circle with a number in it to the Admin bar. The number represents the number of serious issues (errors) you have on your site that need to be addressed immediately. 
* Hovering over the "1" indicator reveals the text "One problem was detected with your Backdrop installation." 
* Clicking on the indicator will take you to the status report page.
![screen shot 2016-01-12 at 1 52 01 am](https://cloud.githubusercontent.com/assets/397895/12260246/a1b8af0e-b8cf-11e5-8ec6-21cfd543e779.png)

We have decided to exclude warnings at this time since we don't yet have a way for those to be dismissed, and we don't want to train users to ignore the indicator. Warnings can be added later.

**Comment by docwilmot** [comment]
Worked on this a bit before I noticed my initial PR was closed. 

- Moved hook invocation to System module, changed hook to `hook_system_alerts`
- made messages dismissable
 - cached all `hook_system_alerts` implementations
 - store active alerts in `$_SESSION` and only display those
 - delete all alerts from `$_SESSION` by clicking the icon
 - modules creating implementations with a link would be responsible for code to dismiss the alert on the link target page or by clearing the entry in `$_SESSION`
- use `watchdog_severity_levels()` for levels allowed
- a form for users to choose which levels are allowed, stored in config

Haven't had the time to go over @jenlampton's PR, but just hope that simpler version doesn't make it more difficult to have a more complete solution later. Seems good from the image and description above though.

**Comment by dboulet** [comment]
Just a friendly reminder to those working on this issue that the Admin bar module is optional, and that not every site will use it. Just a thought: we might want to keep fallbacks in place to show alerts on the page when Admin bar is not available.

**Comment by jenlampton** [comment]
Hm, good point @dboulet thanks for raising that. Do we have other admin-bar alternatives in contrib yet? I have been operating under the assumption that "it's so much better than Toolbar maybe no one will need to invent an alternative." 

We have two paths forward:
1) Assume any contrib solution will be smart enough to also include a status indicator (or users accept the risks if not).
2) Leave something in core that detects if admin bar is enabled or not - and if not duplicate that behavior in another way (I say duplicate because if the contrib alternative also provides an indicator now there will be 2)

I think I prefer the former, now that I think it though...

**Comment by quicksketch** [comment]
The giant red errors on half the admin section have always been over-the-top to me. I'm in favor of just nixing them entirely and leave it to the admin bar if present.

If we implement the admin dashboard (#495), that's another location in which we can place an alert of some kind.

I also like the approach @jenlampton is taking here, where this is not any kind of notifications system at all. It's just a single addition to admin bar for system status report. We can leave this open and continue the work on the more elaborate solutions for notifications, which would replace this bare-bones approach.

**Comment by jenlampton** [comment]
:+1:

**Comment by dboulet** [comment]
> Assume any contrib solution will be smart enough to also include a status indicator (or users accept the risks if not).

Sure, that seems reasonable to me.

Personally, I find the Admin bar to be overkill for small projects, and usually end up replacing it with a custom solution which displays only a couple of links. Wouldn’t be hard to also include the code for the alerts, though.

**Comment by docwilmot** [comment]
Yea, I think @jenlampton's simpler idea is best. I think I got carried away there.

**Comment by quicksketch** [comment]
Some notes on the current @jenlampton PR:

The background color: `#EE3D23` is a color that only exists within our Seven theme. I don't think `admin_bar.css` should adopt that color as it's part of the admin theme color scheme.

The CSS goes wonky when in the collapsed mode:

![selection_039](https://cloud.githubusercontent.com/assets/100206/12311693/cb53920e-ba0d-11e5-9daa-963065d63413.png)

It would be preferable to include a label when in the menu mode like this as well. Something like "Site status". This label would be hidden when displaying across the top on desktop.

To test this I made the files directory read-only on my machine. The error notification did not show up immediately but it did after a cache clear. I think that is okay, we can't be checking for errors constantly. However, then I visited the status report by clicking on the notification. Then I fixed the problem by making the files directory writable again. After reloading the page, the site status report had no errors, but the Admin Bar still had a notification in it. The notification should go away immediately just like the status report fixing immediately. Let's clear the admin bar cache automatically when visiting the status report so that the indicator is updated immediately on this page.

**Comment by jenlampton** [comment]
> The background color: #EE3D23 is a color that only exists within our Seven theme. I don't think admin_bar.css should adopt that color as it's part of the admin theme color scheme.

I stole that from the messages style, actually. I thought messages syle was across all themes?

**Comment by jenlampton** [comment]
I just pushed a new PR with the changes to fix the style when the menu is collapsed.  Adding `Site status` as a label too. I'm leaving the color as-is, per reasons above :)

**Comment by quicksketch** [comment]
This is definitely better. Though I still have a few qualms:

- The positioning of the status report above the search reduces the accessibility of the search box, IMO. We could either A) move site status after the search box or B) move the site status out of the dropdown menu entirely. Though that will definitely have some JavaScript repercussions in our responsive JS code.
- Visiting the status report page still doesn't clear the indicator even if there are no problems remaining.

To solve the later problem, we could a couple of things:

- We could put a check in `admin_bar_preprocess_page()` if the URL was the status report page and then flush the admin bar cache.
- Ideally, system.module would notify admin bar of the change in system status (rather than the admin bar needing to check the URL it was on. This could be done by making a cache entry for `system_site_status`, then checking this cache whenever the site status was being rebuilt. If the list changed, call a hook like `hook_system_site_status_change()`. Then have admin bar implement this hook.

The second solution is probably too elaborate for our timeline, so if it comes down to the first, somewhat hacky solution, then we can settle for that.

**Comment by jenlampton** [comment]
> Visiting the status report page still doesn't clear the indicator even if there are no problems remaining.

I like the cache clear hack on the status report page.  Nice idea!

> The positioning of the status report above the search reduces the accessibility of the search box

Yeah, I see what you mean about that. But I don't think the red dot belongs over there with all that user stuff. How about if we moved the red dot to the right of the menu, rather than to the left of the search box?  Like so:
![screen shot 2016-01-14 at 7 52 20 pm](https://cloud.githubusercontent.com/assets/397895/12345015/606428b4-baf8-11e5-959a-7bc325d08ab3.png)

**Comment by quicksketch** [comment]
Personally, I liked the dot more on the right side, but I do appreciate that the dot is next to the "Reports" menu item, so that when you click on the dot, you can see that you're now within the "Reports" section because it is highlighted (as shown in the screenshot above).

But I think moving it around is something that can be handled later.

This looks good when collapsed onto the mobile version as well:

![selection_043](https://cloud.githubusercontent.com/assets/100206/12347772/e0c5f2a4-bb15-11e5-903b-79632d13a833.png)

I couldn't find any other issues, so let's get this merged and then continue on for more elaborate notifications in the future.

Merged into 1.x for Backdrop 1.3.0. woot!

**Comment by quicksketch** [comment]
Hm, should we make a new issue for notification center? I'd love to keep @sutibun's awesome mockups and @klonos ideas, but having this all mixed together is a bit confusing. What do you say?

For now I'll leave this open, even though the notification is added. Please open new issues for any problems or changes related to it.

**Comment by mikemccaffrey** [comment]
> Hm, should we make a new issue for notification center?

Personally, I'd like to see both these notifications and the system tasks from #1368 combined into a single "action center". The action center could contain all the "actions" that people can take. That would obviously include the traditional ones like "flush caches", "run cron", and "run [system] updates". However, it we could also add "update modules" and such, and have it highlighted if updates are available.

One action/notification I'd like to see added is for syncing staged configuration changes. I keep forgetting to import the new changes to my dev site after I sync things from live, and it would be great to see a reminder that changes need to be addressed.

**Comment by klonos** [comment]
I believe that

1. The indicator/badge should be top-level also on collapsed admin bar (mobile). Having it "hidden" under the "Admin bar" menu item defeats the purpose of what I had in mind when I initially filed this issue:

> ...I really hate it to visit the site status report only to find that there are errors/warning that were sitting there for some time waiting for action (db updates for one).

2. Instead of making assumptions (that either users keep admin bar enabled or the maintainer of the replacement module has accounted for alerts) and removing the page message completely, we should wrap it in a conditional that checks whether admin bar is enabled.

I think that for the dashboard/notification center we should have a separate issue and move all mockups/ideas there. This one is getting too long.

**Comment by sutibun** [comment]
> The indicator/badge should be top-level also on collapsed admin bar (mobile)

Agree. To quote myself from earlier: If you don't have it on the top menu...

> ...you're relying on users hovering over a particular parent menu in order to see the red badge under it. It's a passive alert.... If the status report issue is urgent, there's a chance the user will miss it. 
>
> ... The key to a good notification, IMO, is grabbing user attention when you need it and making sure you train users to expect it in one area, as opposed to multiple places.

Anyway

> I think that for the dashboard/notification center we should have a separate issue and move all mockups/ideas there. This one is getting too long.

^_^)b Agree. This issue page is going all over the place. Let's move discussions to a new issue where it will revolve solely on the dashboard/notification.

**Comment by klonos** [comment]
Let's close this and file separate follow-up issues. I've already created one for the notification badge on mobile (#1665) as well as one for the Dashboard thing (#1666 :imp:)
